### PR TITLE
new nodes and uses a superclass to avoid redundant code

### PR DESCRIPTION
**Pull request by bsubei**
made the equalization and backprojection nodes separate. Also created the brightest-pixel threshold node. Since the topic names are params in the launch file, I was able to create launch files that chain the nodes together (with more nodes, we can check all different combinations)

### CODE REVIEW

**Comment by chicagoedt** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.chicagoedt.org/job/Software_IGVC_Upstream/44/
Test Passed.

**Comment by bsubei** [comment]
don't merge this yet, more nodes on the way

**Comment by chicagoedt** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.chicagoedt.org/job/Software_IGVC_Upstream/45/
Test Passed.

**Comment by bsubei** [comment]
btw i need to clean up most of these nodes because they're using a lot of configs and variables that are common to all of them (but each node is specific and uses only some of them)

**Comment by chicagoedt** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.chicagoedt.org/job/Software_IGVC_Upstream/46/
Test Passed.

**Comment by chicagoedt** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.chicagoedt.org/job/Software_IGVC_Upstream/47/
Test Passed.

**Comment by chicagoedt** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.chicagoedt.org/job/Software_IGVC_Upstream/48/
Test Passed.

**Comment by chicagoedt** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.chicagoedt.org/job/Software_IGVC_Upstream/49/
Test Passed.

**Comment by chicagoedt** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.chicagoedt.org/job/Software_IGVC_Upstream/50/
Test Passed.

**Comment by l0g1x** [comment]
@wfriedl Can you take a look at this since you know the most python?

File: line_detection/src/backproject.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+## @author Basheer Subei
+## @email basheersubei@gmail.com
+#######################################################
+##
+## Gabor Filter example
+##
+###############################################################################
+
+
+class line_detection:
```
Comment (wfriedl) [inline line_detection/src/backproject.py]: (Pep8)[https://www.python.org/dev/peps/pep-0008/#class-names] would encourage LineDetection as the class name and I would encourage you to use the OO syntax "LineDetection(object):"

File: line_detection/src/brightest-pixel.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+                                            final_image)[1]).tostring()
+
+        # publishes image message with line pixels in it
+        self.line_image_pub.publish(final_image_message)
+
+    ## end image_callback()
+
+    def reconfigure_callback(self, config, level):
+
+        # TODO check if the keys exist in the config dictionary or else error
```
Comment (wfriedl) [inline line_detection/src/brightest-pixel.py]: config.get('key_of_intereste', alternate_default_value)

File: line_detection/src/brightest-pixel.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+        brightest_pixel_indices = np.argmax(roi, axis=1)
+        # get the values of max pixels along each row
+        brightest_pixel_values = np.amax(roi, axis=1)
+
+        # make an empty image
+        brightest_pixels = np.zeros(roi.shape)
+
+        count = 0
+        # now fill the image only with the brightest_pixels
+        for index in brightest_pixel_indices:
```
Comment (wfriedl) [inline line_detection/src/brightest-pixel.py:44]:
```
for row, (col, pix) in enumerate(zip(brightest_pixel_indices, brightest_pixel_values):
    brightest_pixels[row, col] = pix
```

Never use a count mechanism when using a for loop; use enumerate.
Use better variable names; especially in loops.
Use the item; versus looking up the item. 
The code will run faster and is easier to figure out!

[Transforming Code into Beautiful, Idiomatic Python](https://www.youtube.com/watch?v=OSGv2VnC0go&index=3&list=PL7xwbQW5nwgw_E5OxYy01xJxzZIUpQxBN)

**Comment by bsubei** [comment]
thanks for the tips, I'll work on these as soon as I'm back in Chicago

**Comment by l0g1x** [comment]
test this please

**Comment by bsubei** [comment]
@wfriedl check it out 7d8b779c5a15581e4ef0815b55a06d17027504f1
I also improved it by using `izip` instead of just `zip`, which uses the iterator and avoids creating another big array (or image)

That video you linked was really good, I need to become better at Python :)

**Pull request merged**